Support parallel Maven builds

First, let me say THANK YOU for such a good plugin. Finally someone made a working protoc maven plugin that behaves as it should.

One minor quirk, though: When running my build in parallel (https://cwiki.apache.org/confluence/display/MAVEN/Parallel+builds+in+Maven+3), Maven complains that:

    [INFO] ------------------------------------------------------------------------
    [INFO] Building XXX 00.00.1-SNAPSHOT
    [INFO] ------------------------------------------------------------------------
    [WARNING] *****************************************************************
    [WARNING] * Your build is requesting parallel execution, but project      *
    [WARNING] * contains the following plugin(s) that are not marked as       *
    [WARNING] * @threadSafe to support parallel building.                     *
    [WARNING] * While this /may/ work fine, please look for plugin updates    *
    [WARNING] * and/or request plugins be made thread-safe.                   *
    [WARNING] * If reporting an issue, report it against the plugin in        *
    [WARNING] * question, not against maven-core                              *
    [WARNING] *****************************************************************
    [WARNING] The following plugins are not marked @threadSafe in XXX:
    [WARNING] com.comoyo.maven.plugins:protoc-bundled-plugin:1.4.47
    [WARNING] *****************************************************************

Could it be made sure that this plugin works well in parallel and if yes, be marked as such, please?

Hi again!

I've been testing this for the past few days as it seemed to work correctly. However, I've seen a failure or two. Basically the results are exactly as you say:

1. Maven 3.2.3: both locally (Windiws 7) and on our build servers (RHEL 6.5), this happens upon the first build:

        [INFO] Using protoc com.comoyo.maven.plugins:protoc-bundled-plugin:exe:2.4.1-linux-x86_64:1.4.47
        [INFO] Using protoc com.comoyo.maven.plugins:protoc-bundled-plugin:exe:2.4.1-linux-x86_64:1.4.47
        [INFO] Downloading: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe
        [INFO] Downloading: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe
        [INFO] Using protoc com.comoyo.maven.plugins:protoc-bundled-plugin:exe:2.4.1-linux-x86_64:1.4.47
        [INFO] Downloading: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe
        [INFO] Downloaded: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe (2497 KB at 2438.2 KB/sec)
        [INFO] Compiling /xxxxxx/xxxxxx/xxxxxx/src/main/protobuf to /xxxxxx/xxxxxx/xxxxxx/target/generated-sources/protobufain]
        [INFO] Downloaded: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe (2497 KB at 2837.2 KB/sec)
        Failed to execute goal com.comoyo.maven.plugins:protoc-bundled-plugin:1.4.47:run (default) on project xxxxxx: Unable to compile /xxxxxx/xxxxxx/xxxxxx/src/main/protobuf/xxxxxx.proto

        [ERROR] Failed to execute goal com.comoyo.maven.plugins:protoc-bundled-plugin:1.4.47:run (default) on project xxxxxx: Unable to compile /xxxxxx/xxxxxx/xxxxxx/src/main/protobuf/xxxxxx.proto: Cannot run program "/xxxxxx/xxxxxx/m2-repository/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe": error=13, Permission denied -> [Help 1]

2. When using Maven 3.2.5 or newer (I tried both 3.3.3 and 3.3.9), I could not make it fail. Simultaneous downloads of protoc seem to work fine, only one thread actually downloads and the others simply wait and then report a successful download of 0 B at 0 KB/s.


Therefore, this is clearly a Maven bug and you don't need to care about it too much (a README notice would be awesome, though). The rest works fine, and therefore this plugin deserves the `@threadSafe` annotation. And 2.6.1 protoc.

Thank you!

--- a/protoc-bundled-plugin/src/main/java/com/comoyo/maven/plugins/protoc/ProtocBundledMojo.java
+++ b/protoc-bundled-plugin/src/main/java/com/comoyo/maven/plugins/protoc/ProtocBundledMojo.java
@@ -180,7 +180,15 @@ public class ProtocBundledMojo extends AbstractMojo
      */
     private File protocExec;
 
-
+    /*
+     * A global static lock to disallow concurrent download and more
+     * importantly concurrent write of the protoc compiler when the plugin
+     * is called multiple times in parallel.
+     * <p>
+     * This is a workaround for a bug in Maven versions <= 3.2.3.
+     */
+    private static final Object protocDownloadLock = new Object();
+    
     private static final Map<String, String> osNamePrefix = new HashMap<String, String>();
     private static final Map<String, String> osArchCanon = new HashMap<String, String>();
     static {
@@ -244,7 +252,7 @@ public class ProtocBundledMojo extends AbstractMojo
      *
      * @param protocName   protoc specifier
      */
-    private synchronized File resolveProtocArtifact(String protocName)
+    private File resolveProtocArtifact(String protocName)
         throws MojoExecutionException
     {
         Artifact artifact
@@ -258,7 +266,10 @@ public class ProtocBundledMojo extends AbstractMojo
         ArtifactResolutionRequest request = new ArtifactResolutionRequest()
             .setArtifact(artifact)
             .setRemoteRepositories(remoteRepositories);
-        ArtifactResolutionResult result = repositorySystem.resolve(request);
+        ArtifactResolutionResult result;
+        synchronized (protocDownloadLock) {
+            result = repositorySystem.resolve(request);
+        }
         if (!result.isSuccess()) {
             throw new MojoExecutionException(
                 "Unable to resolve dependency on protoc binary artifact, sorry: "

